python/ray/data/_internal/execution/streaming_executor.py

```diff
@@ -528,3 +507,36 @@ def _log_op_metrics(topology: Topology) -> None:
     for op in topology:
         log_str += f"{op.name}: {op.metrics.as_dict()}\n"
     logger.debug(log_str)
+
+
+class _ClosingIterator(OutputIterator):
+    """Iterator automatically shutting down executor upon exhausting the
+    iterable sequence.
+
+    NOTE: If this iterator isn't fully exhausted, executor still have to
+          be closed manually by the caller!
```

Comment:
don't we shutdown the executor anyway in `__de__`?

Reply:
Yes, but this will only be called once object is GC'd (which we have no idea when that would happen)